Userland/useradd.cpp

```diff
@@ -0,0 +1,110 @@
+#include <AK/String.h>
+#include <AK/StringBuilder.h>
+#include <LibCore/CArgsParser.h>
+#include <pwd.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+static constexpr uid_t BASE_UID = 1000;
+static constexpr gid_t USERS_GID = 100;
+const char* DEFAULT_SHELL = "/bin/sh";
+
+int main(int argc, char** argv)
+{
+    CArgsParser args_parser("useradd");
+    args_parser.add_arg("d", "HOME_DIR", "home directory for the new user");
+    args_parser.add_arg("u", "UID", "user ID (uid) for the new user");
+    args_parser.add_arg("g", "GROUP", "group ID (gid) for the new user");
+    args_parser.add_arg("m", "create home directory if it does not exist");
+    args_parser.add_arg("s", "SHELL", "name of default shell binary for the new user");
+    args_parser.add_arg("n", "NAME", "gecos name of the new user");
+    CArgsParserResult args = args_parser.parse(argc, argv);
+
+    if (argc <= 1) {
+        args_parser.print_usage();
+        return 1;
+    }
+
+    const char* username = argv[argc - 1]; // Always the user name!
+    uid_t uid = BASE_UID;
+    gid_t gid = USERS_GID;
+
+    // First, let's sort out the uid for the user
+    if (args.is_present("u")) {
+        bool ok;
+        uid = static_cast<uid_t>(args.get("u").to_uint(ok));
+
+        if (getpwuid(uid)) {
+            fprintf(stderr, "uid %u already exists!\n", uid);
+            return 4;
+        }
+
+        if (!ok) {
+            fprintf(stderr, "Invalid uid %s\n", args.get("u").characters());
+            return -1;
+        }
+    } else {
+        uid = BASE_UID;
+        for (; getpwuid(uid); uid++) {
+        }
+    }
+
+    // Now let's sort out the gid!
+    if (args.is_present("g")) {
+        bool ok;
+        gid = static_cast<gid_t>(args.get("g").to_uint(ok));
+
+        if (!ok) {
+            fprintf(stderr, "Invalid gid %s\n", args.get("g").characters());
+            return -1;
+        }
+    }
+
+    AK::String shell = args.is_present("s") ? args.get("s").characters() : DEFAULT_SHELL;
```

Comment:
All of these are imported into the main namespace, so you can just say `String`, etc without the `AK::`

Reply:
Don't turn a string into a characters pointer and then immediately back to a string. Instead, ask the RHS to be converted to a string explicitly.

Reply:
Can you explain this a little bit more? I don't quite get how this would turn into c++. The RHS would be be wrapped in `String()`??

Reply:
Yep,

```cpp
String shell = args.is_preset("s") ? args.get("s") : String(DEFAULT_SHELL);
```

but considering my other comment below, you may not need this, then you'd do

```cpp
const char* shell = args.is_preset("s") ? args.get("s").characters() : DEFAULT_SHELL;
```